Fix handling of 'join' on top of 'union' datasources.

gianm opened the pull request:
The problem is that unions are typically rewritten into a series of individual queries on the underlying tables, but this isn't done when the union is wrapped in a join.

The main changes are in UnionQueryRunner:

1) Replace an `instanceof UnionQueryRunner` check with DataSourceAnalysis.
2) Replace a `query.withDataSource` call with a new function, `Queries.withBaseDataSource`.

Together, these enable UnionQueryRunner to "see through" a join.

a2l007 commented:
We can get rid of UnionQueryRunner itself with  #10030 :)

gianm commented:
> We can get rid of UnionQueryRunner itself with #10030 :)

Ah, that looks interesting. I just commented on it.

abhishekagarwal87 commented on processing/src/main/java/org/apache/druid/query/UnionQueryRunner.java:
```diff
@@ -41,30 +42,27 @@ public UnionQueryRunner(
+
+    final DataSourceAnalysis analysis = DataSourceAnalysis.forDataSource(query.getDataSource());
+
+    if (analysis.isConcreteTableBased() && analysis.getBaseTableDataSources().get().size() != 1) {
```
can we use more explicit condition such as `analysis.getBaseDataSource() instanceof UnionDataSource` than this `analysis.getBaseTableDataSources().get().size() != 1`?  May be the assumption around number of tables will not be valid in future.

gianm replied:
Hmm, I was trying to avoid using `instanceof` checks outside of the DataSourceAnalysis class, and instead stick to cleanly defined interfaces as much as possible. So, I designed dataSourceAnalysis.getBaseTableDataSources() to encapsulate this logic (it has the `instanceof` check you're talking about).

Do you think it'd be better to instead move the logic to UnionQueryRunner?

abhishekagarwal87 replied:
In that case, It would also be fine to have a static method `isUnionDataSource()` within `DataSourceAnalysis` class itself that can be used here as well as in `dataSourceAnalysis.getBaseTableDataSources()`

gianm replied:
Hmm, in thinking about your comment I realized there is a bug here. The code doesn't work right if you have a UnionDataSource of a single TableDataSource. I'll fix it and add a test for this, and while doing so, I'll consider your comments about how to structure the code.

Thanks for the comments!

gianm replied:
I pushed up a new commit that changes the approach a bit; let me know what you think. The `getBaseTableDataSources()` concept is gone and replaced by two separate methods, `getBaseTableDataSource()` and `getBaseUnionDataSource()`.

abhishekagarwal87 replied:
Looks good now.

gianm replied:
Great!

gianm commented:
I just pushed a new commit that changes the approach a bit: I reverted back to the old getBaseTableDataSource() behavior, and added a new method getBaseUnionDataSource(). I also added tests for unions of a single table.

abhishekagarwal87 commented on processing/src/main/java/org/apache/druid/query/UnionQueryRunner.java:
```diff
@@ -41,34 +44,51 @@ public UnionQueryRunner(
+      if (unionDataSource.getDataSources().isEmpty()) {
+        // Shouldn't happen, because UnionDataSource doesn't allow empty unions.
+        throw new ISE("Unexpectedly received empty union");
+      } else if (unionDataSource.getDataSources().size() == 1) {
```
is it optimization because of which this `else if` is not folded into `else`?

gianm replied:
Yeah, the idea is let's not create a MergeSequence if there's only going to be 1 thing.
